dataframes join (#2775)
JuliaData/DataFrames.jl: docs/src/man/joins1.md

# Joining DataFrames

> "I have two different data sets in Julia but I am not sure how to use them. What 

Reviewer: maybe be more precise in "to use them" and vaguely describe what you want to do with them?

criteria should I consider? What are the different ways I can join these data sets?"

Sound familiar? You may have come across this question plenty of times on online discussion
forums. Working with one dataframe is fairly straightforward but things become challenging when

> we have data spread across two or more data frames. This is where the concept of `joins` come in.

Reviewer: Here (and below) - this is not code, but concept, so better use highlight not code block to signal this
```suggestion
we have data spread across two or more data frames. This is where the concept of *joins* come in.
```

> `Joins` is another very common and important operation that arises in the world of tabulated data. 

Reviewer: what do you mean by "another"? Another in comparison to what?

> A join across two DataFrames is the action of combining the two datasets based on shared column values 

Reviewer: ```suggestion
A join across two data frames is the action of combining the two datasets based on shared column values 
```

> that exist across the two tables. We call this column (or columns) the `key`. So, each record from the 

Reviewer: ```suggestion
that exist across them. We call this column (or columns) the *key(s)*. So, each record from the 
```

> first table is matched to a record in the second table — as long as the records’ values are the same. 

Reviewer: ```suggestion
first table is matched to a record in the second table — as long as the records’ key values are the same. 
```

I’m sure you’re quite familiar with e-commerce sites like Amazon and Flipkart these days. We are 
bombarded by their advertisements when we’re visiting non-related websites – that’s the power of 
targeted marketing! Previously, we have used German Credit Card dataset. Now let's say you are working

> on a Classification problem but we have not target column in our dataset. We are given two data frames – 

Reviewer: ```suggestion
on a Classification problem but we do not have target column in our dataset. We are given two data frames – 
```

> on a Classification problem but we have not target column in our dataset. We are given two data frames – 

Reviewer: ```suggestion
on a classification problem but we have not target column in our dataset. We are given two data frames – 
```

> one which contains data about customer and the other that has taget information. We will use these dataframes 

Reviewer: ```suggestion
one which contains data about customer and the other that has taget information. We will use these data frames 
```

to understand how the different types of joins work using `DataFrames.jl`.

For example, suppose that we have the following two datasets:

```jldoctest dataframe
julia> using DataFrames

julia> using CSV

> julia> german_ref = CSV.read((joinpath(dirname(pathof(DataFrames)),
>                                     "..", "docs", "src", "assets", "german.csv")),
>                                     DataFrame)

Reviewer: ```suggestion
julia> german_ref = CSV.read((joinpath(dirname(pathof(DataFrames)),
                                       "..", "docs", "src", "assets", "german.csv")),
                             DataFrame)
```

1000×10 DataFrame
  Row │ id     Age    Sex     Job    Housing  Saving accounts  Checking accoun ⋯
      │ Int64  Int64  String  Int64  String   String           String          ⋯
──────┼─────────────────────────────────────────────────────────────────────────

>     1 │     0     67  male        2  own      NA               little          ⋯

Reviewer: what does `NA` mean? Is this a missing value? then we should properly read it in (the same for the other PR)

    2 │     1     22  female      2  own      little           moderate
    3 │     2     49  male        1  own      little           NA
    4 │     3     45  male        2  free     little           little
    5 │     4     53  male        2  free     little           little          ⋯
    6 │     5     35  male        1  free     NA               NA
    7 │     6     53  male        2  own      quite rich       NA
    8 │     7     35  male        3  rent     little           moderate
  ⋮   │   ⋮      ⋮      ⋮       ⋮       ⋮            ⋮                ⋮        ⋱
  994 │   993     30  male        3  own      little           little          ⋯
  995 │   994     50  male        2  own      NA               NA
  996 │   995     31  female      1  own      little           NA
  997 │   996     40  male        3  own      little           little
  998 │   997     38  male        2  own      little           NA              ⋯
  999 │   998     23  male        2  free     little           little
 1000 │   999     27  male        2  own      moderate         moderate
                                                  4 columns and 985 rows omitted
```

> This table has all the `ids` of all factors to make comparisions for risk factor. Let’s say we 

Reviewer: it is not clear what you mean by "factor". Also why do you earlier discussed some marketing application so why do we talk about risk here? Also why Risk is capitalized?

have another table, which has the Risk of these individuals:

```jldoctest dataframe

> julia> risk_ref = CSV.read((joinpath(dirname(pathof(DataFrames)),
>                                     "..", "docs", "src", "assets", "risk.csv")),
>                                     DataFrame)

Reviewer: ```suggestion
julia> risk_ref = CSV.read(joinpath(dirname(pathof(DataFrames)),
                                    "..", "docs", "src", "assets", "risk.csv"),
                           DataFrame)
```

1000×2 DataFrame
  Row │ id     Risk
      │ Int64  String
──────┼───────────────
    1 │     0  good
    2 │     1  bad
    3 │     2  good
    4 │     3  good
    5 │     4  bad
    6 │     5  good
    7 │     6  good
    8 │     7  good
  ⋮   │   ⋮      ⋮
  994 │   993  good
  995 │   994  good
  996 │   995  good
  997 │   996  good
  998 │   997  good
  999 │   998  bad
 1000 │   999  good
      985 rows omitted
```

We now have 2 tables:
- `german_ref` holds the `id`s, `Age`, `Sex`, `Job` and so on
- `risk_ref` holds the `id`s and `Risk`.

We would like to combine the 2 tables so that we can see columns of `german_ref` and `risk_ref` together. 

> We do a `join`!

Reviewer: maybe be more specific what kind of join?

> Rules for the `on` keyword argument:

Reviewer: ```suggestion
Rules for the `on` keyword argument are as follows:
```

> - a single Symbol or string if joining on one column with the same name, e.g. `on=:id`

Reviewer: ```suggestion
- a single `Symbol` or string if joining on one column with the same name, e.g. `on=:id`
```

> - a Pair of Symbols or string if joining on one column with different names, e.g. `on=:id=>:id2`

Reviewer: ```suggestion
- a `Pair` of `Symbol`s or strings if joining on one column with different names, e.g. `on=:id=>:id2`
```

Reviewer: same comments below

- a vector of Symbols or strings if joining on multiple columns with the same name, 
  e.g. `on=[:id1, :id2]`
- a vector of Pairs of Symbols or strings if joining on multiple columns with the same name, 
  e.g. `on=[:a1=>:a2, :b1=>:b2]`
- a vector containing a combination of Symbols or strings or Pair of Symbols or strings, 
  e.g. `on=[:a1=>:a2, :b1]`

> Here, we are working with a average data set that contains both the `:Age` and `:Sex` for each ID. 

Reviewer: what do you mean by "average"? also ID should be `:id`

    5 │     4     53  male        2  free     little           little          ⋯
    6 │     5     35  male        1  free     NA               NA
    7 │     6     53  male        2  own      quite rich       NA
    8 │     7     35  male        3  rent     little           moderate
  ⋮   │   ⋮      ⋮      ⋮       ⋮       ⋮            ⋮                ⋮        ⋱
  994 │   993     30  male        3  own      little           little          ⋯
  995 │   994     50  male        2  own      NA               NA
  996 │   995     31  female      1  own      little           NA
  997 │   996     40  male        3  own      little           little
  998 │   997     38  male        2  own      little           NA              ⋯
  999 │   998     23  male        2  free     little           little
 1000 │   999     27  male        2  own      moderate         moderate
                                                  5 columns and 985 rows omitted
```

Let’s go through this in detail. Arguments 1 and 2 `german_ref` & `risk_ref` are the two tables we’re 
joining. Argument 3 `on` tells us what the key column is. We’ll use this column to match the observations 

> across the tables. As you can see, we ended up with `1000 rows` and `11 columns`. Now go back to the 

Reviewer: why code escape?

determine which rows should be combined during a join are called keys.

The following functions are provided to perform seven kinds of joins:

- `innerjoin`: the output contains rows for values of the key that exist in all passed data frames.
- `leftjoin`: the output contains rows for values of the key that exist in the first (left) argument, 
  whether or not that value exists in the second (right) argument.
- `rightjoin`: the output contains rows for values of the key that exist in the second (right) argument, 
  whether or not that value exists in the first (left) argument.
- `outerjoin`: the output contains rows for values of the key that exist in any of the passed data frames.
- `semijoin`: Like an inner join, but output is restricted to columns from the first (left) argument.
- `antijoin`: The output contains rows for values of the key that exist in the first (left) but not the 
   second (right) argument. As with semijoin, output is restricted to columns from the first (left) argument.
- `crossjoin`: The output is the cartesian product of rows from all passed data frames.

> See [the Wikipedia page on SQL joins](https://en.wikipedia.org/wiki/Join_(SQL)) for more information.

Reviewer: does it cover all the above listed joins?

> Now to explain further operations I will use random dataset. But fear not, we will now go through all of these 

Reviewer: why it is called "random"?

Here we kept all of the observations from table `persons` no matter what is going on in table `earnings`. 
For records without a match in table `persons`, the salary column has `missingas` value. This makes sense 
as we never actually saw those earning figures. Of course, there is also a `rightjoin`. This keeps all rows 
from the second table.

```jldoctest dataframe
julia> rightjoin(persons, earnings, on = :id)
6×3 DataFrame
 Row │ id     name     salary
     │ Int64  String?  Int64
─────┼────────────────────────
   1 │     3  Rahul      5000
   2 │     4  Vijay      1000
   3 │     5  Akshat     5000
   4 │     6  missing    8000
   5 │     7  missing    3000
   6 │     8  missing    9000
```

If we want to have all ids from both tables, we use an `outerjoin`:

```jldoctest dataframe
julia> outerjoin(persons, earnings, on = :id)
8×3 DataFrame
 Row │ id     name     salary
     │ Int64  String?  Int64?
─────┼─────────────────────────
   1 │     3  Rahul       5000
   2 │     4  Vijay       1000
   3 │     5  Akshat      5000
   4 │     1  Rohit    missing
   5 │     2  Mohit    missing
   6 │     6  missing     8000
   7 │     7  missing     3000
   8 │     8  missing     9000

> julia> antijoin(persons, earnings, on = :id)

Reviewer: why `antijoin` here?

     │ Int64  String  Int64  Int64
─────┼──────────────────────────────
   1 │     1  Rohit       3    5000
   2 │     1  Rohit       4    1000
   3 │     1  Rohit       5    5000
   4 │     1  Rohit       6    8000
   5 │     1  Rohit       7    3000
   6 │     1  Rohit       8    9000
   7 │     2  Mohit       3    5000
   8 │     2  Mohit       4    1000
  ⋮  │   ⋮      ⋮       ⋮      ⋮
  24 │     4  Vijay       8    9000
  25 │     5  Akshat      3    5000
  26 │     5  Akshat      4    1000
  27 │     5  Akshat      5    5000
  28 │     5  Akshat      6    8000
  29 │     5  Akshat      7    3000
  30 │     5  Akshat      8    9000
                     15 rows omitted
```

In the above example, A `crossjoin` took all the rows from `persons` table and for each row, it matches it 
to every single row of `earnings` table. At first impression, this might not make any sense, but it’s a great 
method for finding all combinations of 2 tables. Our new table has 30 rows, which is 
5 (rows of persons) x 6 (rows of earnings).

!!! note

  Cross joins are the only kind of join that does not use a `on` key:
  `crossjoin(persons, earnings, makeunique = true)`

> For more clarification, now let’s say we want to design a new Table by changing the ingredients. To 

Reviewer: maybe give some introduction?

   1 │     1  Rohit
   2 │     2  Mohit
   3 │     3  Rahul
   4 │     4  Vijay
   5 │     5  Akshat

julia> C_names.city = ["Orai", "Gwalior"][mod1.(1:5, 2)]
5-element Vector{String}:
 "Orai"
 "Gwalior"
 "Orai"
 "Gwalior"
 "Orai"

julia> D_earnings = copy(earnings) # do the same for earnings
6×2 DataFrame
 Row │ id     salary
     │ Int64  Int64
─────┼───────────────
   1 │     3    5000
   2 │     4    1000
   3 │     5    5000
   4 │     6    8000
   5 │     7    3000
   6 │     8    9000

julia> D_earnings.city = ["Orai", "Gwalior"][mod1.(1:6, 2)]
6-element Vector{String}:
 "Orai"
 "Gwalior"
 "Orai"
 "Gwalior"
 "Orai"
 "Gwalior"

julia> C_names
5×3 DataFrame
 Row │ id     name    city
     │ Int64  String  String
─────┼────────────────────────
   1 │     1  Rohit   Orai
   2 │     2  Mohit   Gwalior
   3 │     3  Rahul   Orai
   4 │     4  Vijay   Gwalior
   5 │     5  Akshat  Orai

julia> D_earnings
6×3 DataFrame
 Row │ id     salary  city
     │ Int64  Int64   String
─────┼────────────────────────
   1 │     3    5000  Orai
   2 │     4    1000  Gwalior
   3 │     5    5000  Orai
   4 │     6    8000  Gwalior
   5 │     7    3000  Orai
   6 │     8    9000  Gwalior
```

> One way we can think of this is that we have 2 separate databases. One in New York and one in London. 

Reviewer: why New York and London?

With the systems not knowing about each other, they keep track of the users’ ids separately. So the 
name of user 1 in London is not the same as user 1 in New York. Indeed they are different users! So 
when we merge the 2 tables we want to make sure that the names and earnings are not only matched in 
user ids but also on the database name. Let’s do some joining on both columns then:

```jldoctest dataframe
julia> innerjoin(C_names, D_earnings, on = [:id, :city])
3×4 DataFrame
 Row │ id     name    city     salary
     │ Int64  String  String   Int64
─────┼────────────────────────────────
   1 │     3  Rahul   Orai       5000
   2 │     4  Vijay   Gwalior    1000
   3 │     5  Akshat  Orai       5000
```

> As we can see, only Daniel appears in both datasets and lives in the same place.

Reviewer: what is Daniel?

Reviewer: I see you are using https://towardsdatascience.com/joining-dataframes-in-julia-c435e3da32f3. Have you read https://policy.medium.com/medium-terms-of-service-9db0094a1e0f?

Reviewer: I stop reviewing here as we should first ask @niczky12 for permission to reuse.

Reviewer 2: I am happy for this to be reused here. Glad my articles are useful somewhere. 😀

Reviewer: Thank you for giving the permission. That is the spirit I love.

Reviewer 2: If I can, I would like to ask for a reference to the medium article in comments if possible.

Reviewer: This is exactly what I suggested @RohitRathore1 on Slack.
